python/cudf/cudf/core/dataframe.py

```diff
@@ -151,14 +161,35 @@ def __init__(self, data=None, index=None, columns=None):
         # has initializer?
 
         if data is not None:
-            if isinstance(data, dict):
-                data = data.items()
-            elif is_list_like(data) and len(data) > 0:
-                if not isinstance(data[0], (list, tuple)):
+            if is_list_like(data) and len(data) > 0:
+
+                if isinstance(data[0], tuple):
+                    # treat each tuple as a row
+                    def to_col(rows, idx):
+                        return [r[idx] if idx < len(r) else None for r in rows]
+
+                    column_count = max(len(row) for row in data)
+                    data = {
+                        idx: to_col(data, idx) for idx in range(column_count)
+                    }
```

Comment:
Right now it looks like we're figuring out how many columns we'll have and then for each column, iterating through the tuples and picking up the relevant item if possible. Completely playing devils advocate, would it save us some iteration if instead we looked at each row only once and just iterated over it's length, inserting into the relevant column? Might make a difference for wider datasets that have lots of nulls.

Reply:
Probably. But it'd only be saving 1 columns-worth of iteration, and complexity-wise `O(N) == O(M)` where `M=N+!`. I'd rather have the simpler code. If consumers want performance, they should transform the data ahead of time, and utilize the columns-based initialization, IMO.

Reply:
Oh, another reason I did it this way is because we'd have to back-fill the column anyways, which could be a single memcopy, or it could be linear. I don't know, so optimization in that regard seems premature.

Reply:
Instead of doing this manually you can use `itertools.zip_longest` which will handle all of this logic for you and return you a list of tuples representing columns.

Reply:
With the zip_longest iterator you can use a dictionary comprehension to build the dictionary

Reply:
Done. :man_facepalming: